From 41c6413cb54bf338d7a59ed789ec2e0e44c35e6f Mon Sep 17 00:00:00 2001
From: George Wilson <george.wilson@delphix.com>
Date: Wed, 30 Mar 2016 16:32:48 -0700
Subject: [PATCH] 6739 userland version of cv_timedwait_hires() always assumes
 absolute time Reviewed by: Paul Dagnelie <pcd@delphix.com> Reviewed by:
 Matthew Ahrens <mahrens@delphix.com> Reviewed by: Dan McDonald
 <danmcd@omniti.com> Reviewed by: Robert Mustacchi <rm@joyent.com> Approved
 by: Robert Mustacchi <rm@joyent.com>

---
 lib/libzpool/kernel.c          | 7 +++++--
 include/sys/zfs_context.h | 3 ++-
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/lib/libzpool/kernel.c b/lib/libzpool/kernel.c
index 9b8a354..b0f49bf 100644
--- a/lib/libzpool/kernel.c
+++ b/lib/libzpool/kernel.c
@@ -343,10 +343,13 @@ cv_timedwait_hires(kcondvar_t *cv, kmutex_t *mp, hrtime_t tim, hrtime_t res,
 	timestruc_t ts;
 	hrtime_t delta;
 
-	ASSERT(flag == 0);
+	ASSERT(flag == 0 || flag == CALLOUT_FLAG_ABSOLUTE);
 
 top:
-	delta = tim - gethrtime();
+	delta = tim;
+	if (flag & CALLOUT_FLAG_ABSOLUTE)
+		delta -= gethrtime();
+
 	if (delta <= 0)
 		return (-1);
 
diff --git a/include/sys/zfs_context.h b/include/sys/zfs_context.h
index 9e4d8ed..d151dd0 100644
--- a/include/sys/zfs_context.h
+++ b/include/sys/zfs_context.h
@@ -21,7 +21,7 @@
 /*
  * Copyright (c) 2005, 2010, Oracle and/or its affiliates. All rights reserved.
  * Copyright 2011 Nexenta Systems, Inc.  All rights reserved.
- * Copyright (c) 2012, 2014 by Delphix. All rights reserved.
+ * Copyright (c) 2012, 2015 by Delphix. All rights reserved.
  * Copyright (c) 2012, Joyent, Inc. All rights reserved.
  */
 
@@ -279,6 +279,7 @@ extern gid_t *crgetgroups(cred_t *cr);
 typedef cond_t kcondvar_t;
 
 #define	CV_DEFAULT	USYNC_THREAD
+#define	CALLOUT_FLAG_ABSOLUTE	0x2
 
 extern void cv_init(kcondvar_t *cv, char *name, int type, void *arg);
 extern void cv_destroy(kcondvar_t *cv);
-- 
1.8.3.1

